File: 0155-min-stack/0155-min-stack.py
# approach: maintain an adjacent stack that stores the mins at each node
# in the stack

class MinStack(object):

    def __init__(self):
        self.stack = []
        self.mins = []

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        
        if self.stack:
            if val < self.mins[-1]:
                self.mins.append(val)
            else:
                self.mins.append(self.mins[-1])
        else:
            self.mins.append(val)
        
        self.stack.append(val)
        
    def pop(self):
        """
        :rtype: None
        """
        
        self.stack.pop()
        self.mins.pop()

    def top(self):
        """
        :rtype: int
        """
        
        return self.stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        
        return self.mins[-1]

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.java
class MinStack {
    // approach: maintain adjacent stack of mins at each node
    
    private Stack<Integer> stack;
    private Stack<Integer> mins;
    
    public MinStack() {
        this.stack = new Stack<>();
        this.mins = new Stack<>();
    }
    
    public void push(int val) {
        stack.push(val);
        
        if (mins.isEmpty()) {
            mins.push(val);
        } else {
            if (val < mins.peek()) {
                mins.push(val);
            } else {
                mins.push(mins.peek());
            }
        }
    }
    
    public void pop() {
        stack.pop();
        mins.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return mins.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */